// @specify_/mangascraper/lib/functions/getAltTitleByLanguage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function getAltTitleByLanguage(lang, altTitles) {
    const REGEX_LANGUAGES = {
        cn: /[\u4E00-\u9FCC\u3400-\u4DB5\uFA0E\uFA0F\uFA11\uFA13\uFA14\uFA1F\uFA21\uFA23\uFA24\uFA27-\uFA29]|[\ud840-\ud868][\udc00-\udfff]|\ud869[\udc00-\uded6\udf00-\udfff]|[\ud86a-\ud86c][\udc00-\udfff]|\ud86d[\udc00-\udf34\udf40-\udfff]|\ud86e[\udc00-\udc1d]+/g,
        jp: /[ぁ-んァ-ン]+/g,
        en: /[a-zA-Z]+/g,
        kr: /(\S*[\u3131-\u314e|\u314f-\u3163|\uac00-\ud7a3]+\S*)/g,
        vi: /[ÀÁÂÃÈÉÊÌÍÒÓÔÕÙÚĂĐĨŨƠàáâãèéêìíòóôõùúăđĩũơƯĂẠẢẤẦẨẪẬẮẰẲẴẶẸẺẼỀỀỂưăạảấầẩẫậắằẳẵặẹẻẽềềểỄỆỈỊỌỎỐỒỔỖỘỚỜỞỠỢỤỦỨỪễệỉịọỏốồổỗộớờởỡợụủứừỬỮỰỲỴÝỶỸửữựỳỵỷỹ]+/g,
    };
    switch (lang) {
        case 'cn':
            /** Check if string contains CHINESE characters and does NOT contain Japanese characters (hiragana & katakana) */
            return altTitles.filter((title) => title.match(REGEX_LANGUAGES.cn) && !title.match(REGEX_LANGUAGES.jp));
        case 'jp':
            /** Check if string contains Japanese characters (hiragana & katakana)
             * Pretty much a string is in Japanese if it at least contains hiragana and katakana.
             * However, this will only work if the string contains hiragana and katakana, but
             * this should work 99% of the time since those alphabets are used a lot
             */
            return altTitles.filter((title) => title.match(REGEX_LANGUAGES.jp));
        case 'en':
            /** Check if string contains latin characters */
            return altTitles.filter((title) => title.match(REGEX_LANGUAGES.en));
        case 'kr':
            /** Check if string contains Korean characters */
            return altTitles.filter((title) => title.match(REGEX_LANGUAGES.kr));
        case 'vi':
            /** Check if string contains Vietnamese characters */
            return altTitles.filter((title) => title.match(REGEX_LANGUAGES.vi));
        default:
            return null;
    }
}
exports.default = getAltTitleByLanguage;
