// @specify_/mangascraper/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ManganatoGenres = exports.MangakakalotGenres = exports.MangahasuGenres = exports.MangahasuTypes = exports.MangaParkGenres = exports.MangaSeeGenres = exports.ReadMngGenres = exports.MangaBoxGenres = exports.MangaBoxOrderBy = exports.MangaBoxStatus = exports.MangaSeeOrderBy = exports.MangaParkOrderBy = exports.initPuppeteer = exports.MangaBox = exports.MangaPark = exports.MangaSee = exports.Mangahasu = exports.Manganato = exports.Mangakakalot = exports.ReadMng = void 0;
const random_useragent_1 = __importDefault(require("random-useragent"));
var readmng_1 = require("./readmng");
Object.defineProperty(exports, "ReadMng", { enumerable: true, get: function () { return __importDefault(readmng_1).default; } });
var mangakakalot_1 = require("./mangakakalot");
Object.defineProperty(exports, "Mangakakalot", { enumerable: true, get: function () { return __importDefault(mangakakalot_1).default; } });
var manganato_1 = require("./manganato");
Object.defineProperty(exports, "Manganato", { enumerable: true, get: function () { return __importDefault(manganato_1).default; } });
var mangahasu_1 = require("./mangahasu");
Object.defineProperty(exports, "Mangahasu", { enumerable: true, get: function () { return __importDefault(mangahasu_1).default; } });
var mangasee_1 = require("./mangasee");
Object.defineProperty(exports, "MangaSee", { enumerable: true, get: function () { return __importDefault(mangasee_1).default; } });
var mangapark_1 = require("./mangapark");
Object.defineProperty(exports, "MangaPark", { enumerable: true, get: function () { return __importDefault(mangapark_1).default; } });
var mangabox_1 = require("./mangabox");
Object.defineProperty(exports, "MangaBox", { enumerable: true, get: function () { return __importDefault(mangabox_1).default; } });
exports.initPuppeteer = {
    args: [
        '--no-sandbox',
        '--disable-setuid-sandbox',
        '--disable-accelerated-2d-canvas',
        '--no-zygote',
        '--renderer-process-limit=1',
        '--no-first-run',
        '--ignore-certificate-errors',
        '--ignore-certificate-errors-spki-list',
        '--disable-dev-shm-usage',
        '--disable-infobars',
        '--lang=en-US,en',
        '--window-size=1920x1080',
        '--disable-extensions',
        '--disable-gpu',
        `--user-agent=${random_useragent_1.default.getRandom((ua) => ua.osName === 'Windows' && ua.browserName === 'Chrome')}`,
    ],
    ignoreHTTPSErrors: true,
};
var MangaParkOrderBy;
(function (MangaParkOrderBy) {
    MangaParkOrderBy["A-Z"] = "a-z";
    MangaParkOrderBy["latest_updates"] = "update";
    MangaParkOrderBy["rating"] = "rating";
    MangaParkOrderBy["new_manga"] = "create";
    MangaParkOrderBy["most_views"] = "views_a";
})(MangaParkOrderBy = exports.MangaParkOrderBy || (exports.MangaParkOrderBy = {}));
var MangaSeeOrderBy;
(function (MangaSeeOrderBy) {
    MangaSeeOrderBy["A-Z"] = "s";
    MangaSeeOrderBy["latest_updates"] = "lt";
    MangaSeeOrderBy["year_released"] = "y";
    MangaSeeOrderBy["popularity(all_time)"] = "v";
    MangaSeeOrderBy["popularity(monthly)"] = "vm";
})(MangaSeeOrderBy = exports.MangaSeeOrderBy || (exports.MangaSeeOrderBy = {}));
var MangaBoxStatus;
(function (MangaBoxStatus) {
    MangaBoxStatus["completed"] = "end";
    MangaBoxStatus["ongoing"] = "on-going";
    MangaBoxStatus["cancelled"] = "canceled";
    MangaBoxStatus["paused"] = "on-hold";
})(MangaBoxStatus = exports.MangaBoxStatus || (exports.MangaBoxStatus = {}));
var MangaBoxOrderBy;
(function (MangaBoxOrderBy) {
    MangaBoxOrderBy["relevance"] = "";
    MangaBoxOrderBy["latest_updates"] = "latest";
    MangaBoxOrderBy["most_views"] = "views";
    MangaBoxOrderBy["trending"] = "trending";
    MangaBoxOrderBy["rating"] = "rating";
    MangaBoxOrderBy["A-Z"] = "alphabet";
    MangaBoxOrderBy["new_manga"] = "new-manga";
})(MangaBoxOrderBy = exports.MangaBoxOrderBy || (exports.MangaBoxOrderBy = {}));
var MangaBoxGenres;
(function (MangaBoxGenres) {
    MangaBoxGenres["Action"] = "action";
    MangaBoxGenres["Adventure"] = "adventure";
    MangaBoxGenres["Anime"] = "anime";
    MangaBoxGenres["Comedy"] = "comedy";
    MangaBoxGenres["Comic"] = "comic";
    MangaBoxGenres["Cooking"] = "cooking";
    MangaBoxGenres["Detective"] = "detective";
    MangaBoxGenres["Doujinshi"] = "doujinshi";
    MangaBoxGenres["Drama"] = "drama";
    MangaBoxGenres["Fantasy"] = "fantasy";
    MangaBoxGenres["Gender Bender"] = "gender-bender";
    MangaBoxGenres["Harem"] = "harem";
    MangaBoxGenres["Historical"] = "historical";
    MangaBoxGenres["Horror"] = "horror";
    MangaBoxGenres["Josei"] = "josei";
    MangaBoxGenres["Manga"] = "manga";
    MangaBoxGenres["Manhua"] = "manhua";
    MangaBoxGenres["Manhwa"] = "manhwa";
    MangaBoxGenres["Martial Arts"] = "martial-arts";
    MangaBoxGenres["Mature"] = "mature";
    MangaBoxGenres["Mecha"] = "mecha";
    MangaBoxGenres["Mystery"] = "mystery";
    MangaBoxGenres["One shot"] = "one-shot";
    MangaBoxGenres["Psychological"] = "psychological";
    MangaBoxGenres["Romance"] = "romance";
    MangaBoxGenres["School Life"] = "school-life";
    MangaBoxGenres["Sci-fi"] = "sci-fi";
    MangaBoxGenres["Seinen"] = "seinen";
    MangaBoxGenres["Shoujo"] = "shoujo";
    MangaBoxGenres["Shounen"] = "shounen";
    MangaBoxGenres["Slice of Life"] = "slice-of-life";
    MangaBoxGenres["Sports"] = "sports";
    MangaBoxGenres["Supernatural"] = "supernatural";
    MangaBoxGenres["Tragedy"] = "tragedy";
    MangaBoxGenres["Webtoon"] = "webtoon";
})(MangaBoxGenres = exports.MangaBoxGenres || (exports.MangaBoxGenres = {}));
var ReadMngGenres;
(function (ReadMngGenres) {
    ReadMngGenres["Action"] = "2";
    ReadMngGenres["Adventure"] = "4";
    ReadMngGenres["Comedy"] = "5";
    ReadMngGenres["Doujinshi"] = "6";
    ReadMngGenres["Drama"] = "7";
    ReadMngGenres["Ecchi"] = "8";
    ReadMngGenres["Fantasy"] = "9";
    ReadMngGenres["Gender Bender"] = "10";
    ReadMngGenres["Harem"] = "11";
    ReadMngGenres["Historical"] = "12";
    ReadMngGenres["Horror"] = "13";
    ReadMngGenres["Josei"] = "14";
    ReadMngGenres["Lolicon"] = "15";
    ReadMngGenres["Martial Arts"] = "16";
    ReadMngGenres["Mature"] = "17";
    ReadMngGenres["Mecha"] = "18";
    ReadMngGenres["Mystery"] = "19";
    ReadMngGenres["One shot"] = "20";
    ReadMngGenres["Psychological"] = "21";
    ReadMngGenres["Romance"] = "22";
    ReadMngGenres["School Life"] = "23";
    ReadMngGenres["Sci-fi"] = "24";
    ReadMngGenres["Seinen"] = "25";
    ReadMngGenres["Shotacon"] = "26";
    ReadMngGenres["Shoujo"] = "27";
    ReadMngGenres["Shoujo Ai"] = "28";
    ReadMngGenres["Shounen"] = "29";
    ReadMngGenres["Shounen Ai"] = "30";
    ReadMngGenres["Slice of Life"] = "31";
    ReadMngGenres["Smut"] = "32";
    ReadMngGenres["Sports"] = "33";
    ReadMngGenres["Supernatural"] = "34";
    ReadMngGenres["Tragedy"] = "35";
    ReadMngGenres["Yaoi"] = "36";
    ReadMngGenres["Yuri"] = "37";
})(ReadMngGenres = exports.ReadMngGenres || (exports.ReadMngGenres = {}));
var MangaSeeGenres;
(function (MangaSeeGenres) {
    MangaSeeGenres[MangaSeeGenres["Action"] = 0] = "Action";
    MangaSeeGenres[MangaSeeGenres["Adult"] = 1] = "Adult";
    MangaSeeGenres[MangaSeeGenres["Adventure"] = 2] = "Adventure";
    MangaSeeGenres[MangaSeeGenres["Comedy"] = 3] = "Comedy";
    MangaSeeGenres[MangaSeeGenres["Doujinshi"] = 4] = "Doujinshi";
    MangaSeeGenres[MangaSeeGenres["Drama"] = 5] = "Drama";
    MangaSeeGenres[MangaSeeGenres["Ecchi"] = 6] = "Ecchi";
    MangaSeeGenres[MangaSeeGenres["Fantasy"] = 7] = "Fantasy";
    MangaSeeGenres[MangaSeeGenres["Gender Bender"] = 8] = "Gender Bender";
    MangaSeeGenres[MangaSeeGenres["Harem"] = 9] = "Harem";
    MangaSeeGenres[MangaSeeGenres["Hentai"] = 10] = "Hentai";
    MangaSeeGenres[MangaSeeGenres["Historical"] = 11] = "Historical";
    MangaSeeGenres[MangaSeeGenres["Horror"] = 12] = "Horror";
    MangaSeeGenres[MangaSeeGenres["Isekai"] = 13] = "Isekai";
    MangaSeeGenres[MangaSeeGenres["Josei"] = 14] = "Josei";
    MangaSeeGenres[MangaSeeGenres["Lolicon"] = 15] = "Lolicon";
    MangaSeeGenres[MangaSeeGenres["Martial Arts"] = 16] = "Martial Arts";
    MangaSeeGenres[MangaSeeGenres["Mature"] = 17] = "Mature";
    MangaSeeGenres[MangaSeeGenres["Mecha"] = 18] = "Mecha";
    MangaSeeGenres[MangaSeeGenres["Mystery"] = 19] = "Mystery";
    MangaSeeGenres[MangaSeeGenres["Psychological"] = 20] = "Psychological";
    MangaSeeGenres[MangaSeeGenres["Romance"] = 21] = "Romance";
    MangaSeeGenres[MangaSeeGenres["School Life"] = 22] = "School Life";
    MangaSeeGenres[MangaSeeGenres["Sci-fi"] = 23] = "Sci-fi";
    MangaSeeGenres[MangaSeeGenres["Seinen"] = 24] = "Seinen";
    MangaSeeGenres[MangaSeeGenres["Seinen Supernatural"] = 25] = "Seinen Supernatural";
    MangaSeeGenres[MangaSeeGenres["Shotacon"] = 26] = "Shotacon";
    MangaSeeGenres[MangaSeeGenres["Shoujo"] = 27] = "Shoujo";
    MangaSeeGenres[MangaSeeGenres["Shoujo Ai"] = 28] = "Shoujo Ai";
    MangaSeeGenres[MangaSeeGenres["Shounen"] = 29] = "Shounen";
    MangaSeeGenres[MangaSeeGenres["Shounen Ai"] = 30] = "Shounen Ai";
    MangaSeeGenres[MangaSeeGenres["Slice of Life"] = 31] = "Slice of Life";
    MangaSeeGenres[MangaSeeGenres["Smut"] = 32] = "Smut";
    MangaSeeGenres[MangaSeeGenres["Sports"] = 33] = "Sports";
    MangaSeeGenres[MangaSeeGenres["Supernatural"] = 34] = "Supernatural";
    MangaSeeGenres[MangaSeeGenres["Tragedy"] = 35] = "Tragedy";
    MangaSeeGenres[MangaSeeGenres["Yaoi"] = 36] = "Yaoi";
    MangaSeeGenres[MangaSeeGenres["Yuri"] = 37] = "Yuri";
})(MangaSeeGenres = exports.MangaSeeGenres || (exports.MangaSeeGenres = {}));
var MangaParkGenres;
(function (MangaParkGenres) {
    MangaParkGenres["4 koma"] = "4-koma";
    MangaParkGenres["Action"] = "action";
    MangaParkGenres["Adaptation"] = "adaptation";
    MangaParkGenres["Adult"] = "adult";
    MangaParkGenres["Adventure"] = "adventure";
    MangaParkGenres["Aliens"] = "aliens";
    MangaParkGenres["Animals"] = "animals";
    MangaParkGenres["Anthology"] = "anthology";
    MangaParkGenres["Award winning"] = "award-winning";
    MangaParkGenres["Comedy"] = "comedy";
    MangaParkGenres["Cooking"] = "cooking";
    MangaParkGenres["Crime"] = "crime";
    MangaParkGenres["Crossdressing"] = "crossdressing";
    MangaParkGenres["Delinquents"] = "delinquents";
    MangaParkGenres["Demons"] = "demons";
    MangaParkGenres["Doujinshi"] = "doujinshi";
    MangaParkGenres["Drama"] = "drama";
    MangaParkGenres["Ecchi"] = "ecchi";
    MangaParkGenres["Fan colored"] = "fan-colored";
    MangaParkGenres["Fantasy"] = "fantasy";
    MangaParkGenres["Food"] = "food";
    MangaParkGenres["Full color"] = "full-color";
    MangaParkGenres["Game"] = "game";
    MangaParkGenres["Gender bender"] = "gender-bender";
    MangaParkGenres["Genderswap"] = "genderswap";
    MangaParkGenres["Ghosts"] = "ghosts";
    MangaParkGenres["Gore"] = "gore";
    MangaParkGenres["Gossip"] = "gossip";
    MangaParkGenres["Gyaru"] = "gyaru";
    MangaParkGenres["Harem"] = "harem";
    MangaParkGenres["Historical"] = "historical";
    MangaParkGenres["Horror"] = "horror";
    MangaParkGenres["Incest"] = "incest";
    MangaParkGenres["Isekai"] = "isekai";
    MangaParkGenres["Josei"] = "josei";
    MangaParkGenres["Kids"] = "kids";
    MangaParkGenres["Loli"] = "loli";
    MangaParkGenres["Lolicon"] = "lolicon";
    MangaParkGenres["Long strip"] = "long-strip";
    MangaParkGenres["Mafia"] = "mafia";
    MangaParkGenres["Magic"] = "magic";
    MangaParkGenres["Magical girls"] = "magical-girls";
    MangaParkGenres["Manhwa"] = "manhwa";
    MangaParkGenres["Martial arts"] = "martial-arts";
    MangaParkGenres["Mature"] = "mature";
    MangaParkGenres["Mecha"] = "mecha";
    MangaParkGenres["Medical"] = "medical";
    MangaParkGenres["Military"] = "military";
    MangaParkGenres["Monster girls"] = "monster-girls";
    MangaParkGenres["Monsters"] = "monsters";
    MangaParkGenres["Music"] = "music";
    MangaParkGenres["Mystery"] = "mystery";
    MangaParkGenres["Ninja"] = "ninja";
    MangaParkGenres["Office workers"] = "office-workers";
    MangaParkGenres["Official colored"] = "official-colored";
    MangaParkGenres["One shot"] = "one-shot";
    MangaParkGenres["Parody"] = "parody";
    MangaParkGenres["Philosophical"] = "philosophical";
    MangaParkGenres["Police"] = "police";
    MangaParkGenres["Post apocalyptic"] = "post-apocalyptic";
    MangaParkGenres["Psychological"] = "psychological";
    MangaParkGenres["Reincarnation"] = "reincarnation";
    MangaParkGenres["Reverse harem"] = "reverse-harem";
    MangaParkGenres["Romance"] = "romance";
    MangaParkGenres["Samurai"] = "samurai";
    MangaParkGenres["School life"] = "school-life";
    MangaParkGenres["Sci fi"] = "sci-fi";
    MangaParkGenres["Seinen"] = "seinen";
    MangaParkGenres["Shota"] = "shota";
    MangaParkGenres["Shotacon"] = "shotacon";
    MangaParkGenres["Shoujo"] = "shoujo";
    MangaParkGenres["Shoujo ai"] = "shoujo-ai";
    MangaParkGenres["Shounen"] = "shounen";
    MangaParkGenres["Shounen ai"] = "shounen-ai";
    MangaParkGenres["Slice of life"] = "slice-of-life";
    MangaParkGenres["Smut"] = "smut";
    MangaParkGenres["Space"] = "space";
    MangaParkGenres["Sports"] = "sports";
    MangaParkGenres["Super power"] = "super-power";
    MangaParkGenres["Superhero"] = "superhero";
    MangaParkGenres["Supernatural"] = "supernatural";
    MangaParkGenres["Survival"] = "survival";
    MangaParkGenres["Suspense"] = "suspense";
    MangaParkGenres["Thriller"] = "thriller";
    MangaParkGenres["Time travel"] = "time-travel";
    MangaParkGenres["Toomics"] = "toomics";
    MangaParkGenres["Traditional games"] = "traditional-games";
    MangaParkGenres["Tragedy"] = "tragedy";
    MangaParkGenres["User created"] = "user-created";
    MangaParkGenres["Vampire"] = "vampire";
    MangaParkGenres["Vampires"] = "vampires";
    MangaParkGenres["Video games"] = "video-games";
    MangaParkGenres["Villainess"] = "villainess";
    MangaParkGenres["Virtual reality"] = "virtual-reality";
    MangaParkGenres["Web comic"] = "web-comic";
    MangaParkGenres["Webtoon"] = "webtoon";
    MangaParkGenres["Wuxia"] = "wuxia";
    MangaParkGenres["Yaoi"] = "yaoi";
    MangaParkGenres["Yuri"] = "yuri";
    MangaParkGenres["Zombies"] = "zombies";
})(MangaParkGenres = exports.MangaParkGenres || (exports.MangaParkGenres = {}));
var MangahasuTypes;
(function (MangahasuTypes) {
    MangahasuTypes["manga"] = "10";
    MangahasuTypes["manhwa"] = "12";
    MangahasuTypes["manhua"] = "19";
})(MangahasuTypes = exports.MangahasuTypes || (exports.MangahasuTypes = {}));
var MangahasuGenres;
(function (MangahasuGenres) {
    MangahasuGenres["4-koma"] = "46";
    MangahasuGenres["Action"] = "1";
    MangahasuGenres["Adaptation"] = "101";
    MangahasuGenres["Adult"] = "2";
    MangahasuGenres["Adventure"] = "3";
    MangahasuGenres["Aliens"] = "103";
    MangahasuGenres["Animals"] = "73";
    MangahasuGenres["Anime"] = "57";
    MangahasuGenres["Anthology"] = "99";
    MangahasuGenres["Award Winning"] = "48";
    MangahasuGenres["Bara"] = "60";
    MangahasuGenres["Comedy"] = "4";
    MangahasuGenres["Comic"] = "5";
    MangahasuGenres["Cooking"] = "6";
    MangahasuGenres["Crime"] = "92";
    MangahasuGenres["Crossdressing"] = "86";
    MangahasuGenres["Delinquents"] = "83";
    MangahasuGenres["Demons"] = "51";
    MangahasuGenres["Doujinshi"] = "7";
    MangahasuGenres["Drama"] = "8";
    MangahasuGenres["Ecchi"] = "9";
    MangahasuGenres["Fan Colored"] = "107";
    MangahasuGenres["Fantasy"] = "10";
    MangahasuGenres["Full Color"] = "95";
    MangahasuGenres["Game"] = "68";
    MangahasuGenres["Gender Bender"] = "11";
    MangahasuGenres["Genderswap"] = "81";
    MangahasuGenres["Ghosts"] = "90";
    MangahasuGenres["Gore"] = "100";
    MangahasuGenres["Gyaru"] = "97";
    MangahasuGenres["Harem"] = "12";
    MangahasuGenres["Historical"] = "13";
    MangahasuGenres["Horror"] = "14";
    MangahasuGenres["Incest"] = "84";
    MangahasuGenres["Isekai"] = "67";
    MangahasuGenres["Josei"] = "15";
    MangahasuGenres["Live Action"] = "59";
    MangahasuGenres["Loli"] = "91";
    MangahasuGenres["Lolicon"] = "16";
    MangahasuGenres["Long Strip"] = "93";
    MangahasuGenres["Mafia"] = "113";
    MangahasuGenres["Magic"] = "55";
    MangahasuGenres["Magical Girls"] = "89";
    MangahasuGenres["Manga Reviews"] = "64";
    MangahasuGenres["Martial Arts"] = "20";
    MangahasuGenres["Mature"] = "21";
    MangahasuGenres["Mecha"] = "22";
    MangahasuGenres["Medical"] = "23";
    MangahasuGenres["Military"] = "62";
    MangahasuGenres["Monster Girls"] = "87";
    MangahasuGenres["Monsters"] = "72";
    MangahasuGenres["Music"] = "24";
    MangahasuGenres["Mystery"] = "25";
    MangahasuGenres["Netorare/NTR"] = "123";
    MangahasuGenres["Ninja"] = "112";
    MangahasuGenres["Office"] = "119";
    MangahasuGenres["Office Workers"] = "80";
    MangahasuGenres["Official Colored"] = "96";
    MangahasuGenres["One shot"] = "26";
    MangahasuGenres["Others"] = "114";
    MangahasuGenres["Philosophical"] = "110";
    MangahasuGenres["Police"] = "105";
    MangahasuGenres["Post-Apocalyptic"] = "76";
    MangahasuGenres["Psychological"] = "27";
    MangahasuGenres["Reincarnation"] = "74";
    MangahasuGenres["Reverse harem"] = "69";
    MangahasuGenres["Romance"] = "28";
    MangahasuGenres["Samurai"] = "108";
    MangahasuGenres["school"] = "118";
    MangahasuGenres["School Life"] = "29";
    MangahasuGenres["School Life. Seinen"] = "115";
    MangahasuGenres["Sci-fi"] = "30";
    MangahasuGenres["Seinen"] = "31";
    MangahasuGenres["Seinen  Supernatural"] = "66";
    MangahasuGenres["Sexual Violence"] = "98";
    MangahasuGenres["Shota"] = "104";
    MangahasuGenres["Shotacon"] = "32";
    MangahasuGenres["Shoujo"] = "33";
    MangahasuGenres["Shoujo Ai"] = "34";
    MangahasuGenres["Shoujoai"] = "63";
    MangahasuGenres["Shounen"] = "35";
    MangahasuGenres["Shounen Ai"] = "36";
    MangahasuGenres["Shounenai"] = "61";
    MangahasuGenres["Slice of Life"] = "37";
    MangahasuGenres["SM/BDSM"] = "122";
    MangahasuGenres["Smut"] = "38";
    MangahasuGenres["Sports"] = "39";
    MangahasuGenres["Super power"] = "70";
    MangahasuGenres["Superhero"] = "88";
    MangahasuGenres["Supernatural"] = "40";
    MangahasuGenres["Survival"] = "77";
    MangahasuGenres["Thriller"] = "75";
    MangahasuGenres["Time Travel"] = "78";
    MangahasuGenres["Traditional Games"] = "111";
    MangahasuGenres["Tragedy"] = "41";
    MangahasuGenres["Uncategorized"] = "65";
    MangahasuGenres["User Created"] = "102";
    MangahasuGenres["Vampire"] = "58";
    MangahasuGenres["Vampires"] = "82";
    MangahasuGenres["Video Games"] = "85";
    MangahasuGenres["Villainess"] = "116";
    MangahasuGenres["Violence"] = "120";
    MangahasuGenres["Virtual Reality"] = "109";
    MangahasuGenres["Web Comic"] = "94";
    MangahasuGenres["Webtoon"] = "42";
    MangahasuGenres["Webtoons"] = "56";
    MangahasuGenres["Western"] = "121";
    MangahasuGenres["Wuxia"] = "71";
    MangahasuGenres["Yaoi"] = "43";
    MangahasuGenres["Youkai"] = "106";
    MangahasuGenres["Yuri"] = "44";
    MangahasuGenres["Zombies"] = "79";
})(MangahasuGenres = exports.MangahasuGenres || (exports.MangahasuGenres = {}));
var MangakakalotGenres;
(function (MangakakalotGenres) {
    MangakakalotGenres["Action"] = "2";
    MangakakalotGenres["Adult"] = "3";
    MangakakalotGenres["Adventure"] = "4";
    MangakakalotGenres["Comedy"] = "6";
    MangakakalotGenres["Cooking"] = "7";
    MangakakalotGenres["Doujinshi"] = "9";
    MangakakalotGenres["Drama"] = "10";
    MangakakalotGenres["Ecchi"] = "11";
    MangakakalotGenres["Fantasy"] = "12";
    MangakakalotGenres["Gender bender"] = "13";
    MangakakalotGenres["Harem"] = "14";
    MangakakalotGenres["Historical"] = "15";
    MangakakalotGenres["Horror"] = "16";
    MangakakalotGenres["Isekai"] = "45";
    MangakakalotGenres["Josei"] = "17";
    MangakakalotGenres["Manhua"] = "44";
    MangakakalotGenres["Manhwa"] = "43";
    MangakakalotGenres["Martial arts"] = "19";
    MangakakalotGenres["Mature"] = "20";
    MangakakalotGenres["Mecha"] = "21";
    MangakakalotGenres["Medical"] = "22";
    MangakakalotGenres["Mystery"] = "24";
    MangakakalotGenres["One shot"] = "25";
    MangakakalotGenres["Psychological"] = "26";
    MangakakalotGenres["Romance"] = "27";
    MangakakalotGenres["School life"] = "28";
    MangakakalotGenres["Sci fi"] = "29";
    MangakakalotGenres["Seinen"] = "30";
    MangakakalotGenres["Shoujo"] = "31";
    MangakakalotGenres["Shoujo ai"] = "32";
    MangakakalotGenres["Shounen"] = "33";
    MangakakalotGenres["Shounen ai"] = "34";
    MangakakalotGenres["Slice of life"] = "35";
    MangakakalotGenres["Smut"] = "36";
    MangakakalotGenres["Sports"] = "37";
    MangakakalotGenres["Supernatural"] = "38";
    MangakakalotGenres["Tragedy"] = "39";
    MangakakalotGenres["Webtoons"] = "40";
    MangakakalotGenres["Yaoi"] = "41";
    MangakakalotGenres["Yuri"] = "42";
})(MangakakalotGenres = exports.MangakakalotGenres || (exports.MangakakalotGenres = {}));
var ManganatoGenres;
(function (ManganatoGenres) {
    ManganatoGenres["Action"] = "2";
    ManganatoGenres["Adult"] = "3";
    ManganatoGenres["Adventure"] = "4";
    ManganatoGenres["Comedy"] = "6";
    ManganatoGenres["Cooking"] = "7";
    ManganatoGenres["Doujinshi"] = "9";
    ManganatoGenres["Drama"] = "10";
    ManganatoGenres["Ecchi"] = "11";
    ManganatoGenres["Fantasy"] = "12";
    ManganatoGenres["Gender bender"] = "13";
    ManganatoGenres["Harem"] = "14";
    ManganatoGenres["Historical"] = "15";
    ManganatoGenres["Horror"] = "16";
    ManganatoGenres["Isekai"] = "45";
    ManganatoGenres["Josei"] = "17";
    ManganatoGenres["Manhua"] = "44";
    ManganatoGenres["Manhwa"] = "43";
    ManganatoGenres["Martial arts"] = "19";
    ManganatoGenres["Mature"] = "20";
    ManganatoGenres["Mecha"] = "21";
    ManganatoGenres["Medical"] = "22";
    ManganatoGenres["Mystery"] = "24";
    ManganatoGenres["One shot"] = "25";
    ManganatoGenres["Psychological"] = "26";
    ManganatoGenres["Romance"] = "27";
    ManganatoGenres["School life"] = "28";
    ManganatoGenres["Sci fi"] = "29";
    ManganatoGenres["Seinen"] = "30";
    ManganatoGenres["Shoujo"] = "31";
    ManganatoGenres["Shoujo ai"] = "32";
    ManganatoGenres["Shounen"] = "33";
    ManganatoGenres["Shounen ai"] = "34";
    ManganatoGenres["Slice of life"] = "35";
    ManganatoGenres["Smut"] = "36";
    ManganatoGenres["Sports"] = "37";
    ManganatoGenres["Supernatural"] = "38";
    ManganatoGenres["Tragedy"] = "39";
    ManganatoGenres["Webtoons"] = "40";
    ManganatoGenres["Yaoi"] = "41";
    ManganatoGenres["Yuri"] = "42";
})(ManganatoGenres = exports.ManganatoGenres || (exports.ManganatoGenres = {}));
